File: leetcode/string/valid_palindrome.cc
// https://leetcode.com/problems/valid-palindrome/

// time complexity: O(n)

#include <string>

using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        int len = s.length();
        int l = 0, r = len - 1;
        while (l < r) {
            while (l < r && !isalnum(s[l]))
                l++;
            while (l < r && !isalnum(s[r]))
                r--;
            if (tolower(s[l++]) != tolower(s[r--]))
                return false;
        }
        return true;
    }
};

File: algorithms/array/valid_palindrome.py
# https://leetcode.com/problems/valid-palindrome/

# time complexity: O(n)

import re

class Solution:
    def isPalindrome(self, s: str) -> bool:
        return (s := re.sub('[^a-z0-9]', '', s.lower())) == s[::-1]